python/lsst/ci/tsort.py

```diff
@@ -1,59 +1,103 @@
-#!/usr/bin/env python
-# -- mjuric: Taken from https://pypi.python.org/pypi/tsort and slightly modified
-# -- mjuric: LGPL licenced
-# -*- coding: utf-8 -*-
-
+from typing import Iterator, Iterable, List, Mapping, Set, Tuple
 
 class GraphError(Exception):
     pass
 
 
-def tsort(edges):
-    """topologically sort vertices in edges.
-    edges: list of pairs of vertices. Edges must form a DAG.
-           If the graph has a cycle, then GraphError is raised.
-    returns: topologically sorted list of vertices.
-    see http://en.wikipedia.org/wiki/Topological_sorting
+def to_dep_graph(edges: Iterable[Tuple[str, str]]) -> Mapping[str, Set[str]]:
+    """Takes an iterable collection of (object, dependency) pairs and returns
+    as a graph set"""
+    graph = {}
+    for node, dep in edges:
+        graph.setdefault(node, set()).add(dep)
+        # Ensure all deps are added to graph	
+        graph.setdefault(dep, set())
+    return graph
+
+
+def toposort(edges: Iterable[Tuple[str, str]]) -> Iterator[List]:
+    """Takes an iterable collection of (object, dependency) pairs
+    and returns an iterator of ordered dependency lists.
+    The items in each list can be processed in any order.
+    """
+    return toposort_mapping(to_dep_graph(edges))
```

Comment:
What do you gain by having the implementation in `toposort_mapping`? Is that a useful method in its own right? If it's not being used (except here) then please consider putting the body of the function here and have the first line be graph_set = to_dep_graph(edges).

Likewise for the dfs variant.

Reply:
Simplified the code and moved the _mapping variants to the named variants.